isolate, deploy: Use Fabric library to run shell commands remotely

lucas-akira opened the pull request:
This fixes an IO desynchronization issue when sending remote sudo commands that could happen if the device has Torizon OS 7.3.0 or higher installed: the password would be sent to the device before sudo is ready to receive it, which would cause sudo to hang waiting for the password until its default timeout (5 minutes) gets reached and fail.

This wasn't reproducible on previous Torizon OS 7 versions. 7.3.0 introduced a newer version of sudo (1.9.17p1), which could explain this difference in behavior: probably this version takes a bit longer to show the password prompt, just enough for the issue to show up.

One way avoid this problem is to send shell commands remotely using a library designed to do this task, like Fabric. It can detect the sudo password prompt and auto-respond to it accordingly, which is a more robust method than the previous one.

-----

Also included is a fix to `isolate` when the device `/etc` has files or directories with leading or trailing whitespaces in their names. This was found in our CI pipeline while testing the fix above.

Closes TCB-729

lucas-akira committed:
isolate, deploy: Use Fabric library to run shell commands remotely

This fixes an IO desynchronization issue when sending remote sudo
commands that could happen if the device has Torizon OS 7.3.0 or higher
installed: the password would be sent to the device before sudo is ready
to receive it, which would cause sudo to hang waiting for the password
until its default timeout (5 minutes) gets reached and fail.

This wasn't reproducible on previous Torizon OS 7 versions. 7.3.0
introduced a newer version of sudo (1.9.17p1), which could explain
this difference in behavior: probably this version takes a bit longer to
show the password prompt, just enough for the issue to show up.

One way avoid this problem is to send shell commands remotely using a
library designed to do this task, like Fabric. It can detect the sudo
password prompt and auto-respond to it accordingly, which is a more
robust method than the previous one.

Related-to: TCB-729

Signed-off-by: Lucas Akira Morishita <lucas.morishita@toradex.com>

lucas-akira committed:
backend/isolate: Support filenames with leading or trailing spaces

This fixes a (very) particular case when the device /etc has files or
directories with leading or trailing whitespaces in their names.

Now any whitespace in .tcattr filename entries is replaced by a
backslash followed by its respective 3-digit octal ASCII code (\040).

Previously the resulting .tcattr had filenames without any changes i.e.
with the exact number of whitespaces. However when running the union
command, setfacl ignores any leading/trailing whitespaces in the
file/directory name entry, resulting in it looking for filenames
without them.

This was originally detected in our test pipeline, where the device
under test had files with this name pattern in /etc, probably as a
result of a previous 'isolate' test suite that created them, but never
deleted them as the it failed and the cleanup routine ended up not
being executed.

Signed-off-by: Lucas Akira Morishita <lucas.morishita@toradex.com>

jsrc27 commented:
Looks good on my side for the most part.

Just one question. Have you tested this with older versions of Torizon OS? From before we had the new `sudo` version, especially on Torizon 6.

I assume there would be no issue since this seems pretty agnostic from those details, but just wanted to be sure.

lucas-akira commented:
I just tested both `isolate` and `deploy` with Torizon OS 6.8.0 and 7.2.0, and they worked as expected.

rborn-tx commented:
@lucas-akira Nice work!

Question:

Have you considered changing `backend/images.py::get_device_info()` to also use Fabric? Since you already did lots of work, converting that function seems relatively simple. If we did that, we could drop `paramiko` rather than have two libraries being used with the same purpose, which would be good for maintenance and consistency.

lucas-akira commented:
Yeah, I considered replacing Paramiko with Fabric in `backend/images.py` as well, but since it isn't explicitly sending any shell commands to be executed on the device side like with `isolate` and `deploy`, and there are no reported issues with the current implementation, I thought that altering it may be out of scope for this PR.

But you've raised a good point: it would be better for code maintenance and consistency using only one library to handle SSH connections to remote devices.

Fabric does use Paramiko under the hood, so TorizonCore Builder would still technically use it, but this doesn't really matter for code maintenance/consistency.

@rborn-tx I'm inclined to follow your suggestion. Any thoughts @jsrc27 ?

jsrc27 commented:
> Any thoughts

I don't have any strong opinions on this as long as nothing changes for the user.

lucas-akira committed:
images download: Replace Paramiko with Fabric to get device info

Since Fabric is already being used for other commands that require an
SSH connection to a device, use it in 'images download' as well for
consistency and better code maintenance.

Related-to: TCB-729

Signed-off-by: Lucas Akira Morishita <lucas.morishita@toradex.com>

rborn-tx reviewed (approved):
Awesome!
